File: 004/triangle.rb
def check(a, b, c)
  (a + b > c) || (a + c > b) || (b + c > a)
end

def type(a, b, c)
  return "Tam giac deu" if a == b && b == c

  return "Tam giac can" if a == b || a == c || b == c

  return "Tam giac vuong" if a**2 + b**2 == c**2 || a**2 + c**2 == b**2 || b**2 + c**2 == a**2

  "Tam giac thuong"
end

def Heron(a, b, c)
  p = (a + b + c)/2.0
  Math.sqrt(p*(p - a)*(p - b)*(p - c))
end

File: 004/triangle.c
#include "triangle.h"

bool check(double a, double b, double c) {
    return (a + b > c) && (a + c > b) && (b + c > a);
}

const char* type(double a, double b, double c) {
    if (!check(a, b, c)) {
        return "Khong phai tam giac";
    }

    if (a == b && b == c) {
        return "Tam giac deu";
    }

    if (a == b || a == c || b == c) {
        return "Tam giac can";
    }

    if (fabs(pow(a, 2) + pow(b, 2) - pow(c, 2)) < 1e-6 ||
        fabs(pow(a, 2) + pow(c, 2) - pow(b, 2)) < 1e-6 ||
        fabs(pow(c, 2) + pow(b, 2) - pow(a, 2)) < 1e-6) {
        return "Tam giac vuong";
    }

    return "Tam giac thuong";
}

double Heron(double a, double b, double c) {
    if (!check(a, b, c)) {
        return -1;
    }
    double p = (a + b + c) / 2;
    return sqrt(p * (p - a) * (p - b) * (p - c));
}

